transfer/custom.go

```diff
@@ -0,0 +1,382 @@
+func (a *customAdapter) abortWorkerProcess(ctx *customAdapterWorkerContext) {
+	tracerx.Printf("xfer: Aborting worker process: %d", ctx.workerNum)
+	ctx.stdin.Close()
+	ctx.stdout.Close()
+	ctx.cmd.Process.Kill()
+}
+func (a *customAdapter) WorkerEnding(workerNum int, ctx interface{}) {
+	customCtx, ok := ctx.(*customAdapterWorkerContext)
+	if !ok {
+		tracerx.Printf("Context object for custom transfer %q was of the wrong type", a.name)
+		return
+	}
+
+	err := a.shutdownWorkerProcess(customCtx)
+	if err != nil {
+		tracerx.Printf("xfer: error finishing up custom transfer process %q worker %d, aborting: %v", a.path, customCtx.workerNum, err)
+		a.abortWorkerProcess(customCtx)
+	}
+}
+
+func (a *customAdapter) DoTransfer(ctx interface{}, t *Transfer, cb TransferProgressCallback, authOkFunc func()) error {
+	if ctx == nil {
+		return fmt.Errorf("Custom transfer %q was not properly initialized, see previous errors", a.name)
+	}
+
+	customCtx, ok := ctx.(*customAdapterWorkerContext)
+	if !ok {
+		return fmt.Errorf("Context object for custom transfer %q was of the wrong type", a.name)
+	}
+	var authCalled bool
+
+	rel, ok := t.Object.Rel(a.getOperationName())
+	if !ok {
+		return errors.New("Object not found on the server.")
+	}
+	var req *customAdapterTransferRequest
+	if a.direction == Upload {
+		req = NewCustomAdapterUploadRequest(t.Object.Oid, t.Object.Size, t.Path, rel)
+	} else {
+		req = NewCustomAdapterDownloadRequest(t.Object.Oid, t.Object.Size, rel)
+	}
+	err := a.sendMessage(customCtx, req)
+	if err != nil {
+		return err
+	}
+
+	// 1..N replies (including progress & one of download / upload)
+	var complete bool
+	for !complete {
+		resp, err := a.readResponse(customCtx)
+		if err != nil {
+			return err
+		}
+		var wasAuthOk bool
+		switch resp.Id {
+		case "progress":
+			// Progress
+			if resp.Oid != t.Object.Oid {
+				return fmt.Errorf("Unexpected oid %q in response, expecting %q", resp.Oid, t.Object.Oid)
+			}
+			if cb != nil {
+				cb(t.Name, t.Object.Size, resp.BytesSoFar, resp.BytesSinceLast)
+			}
+			wasAuthOk = resp.BytesSoFar > 0
+		case "complete":
+			// Download/Upload complete
+			if resp.Oid != t.Object.Oid {
+				return fmt.Errorf("Unexpected oid %q in response, expecting %q", resp.Oid, t.Object.Oid)
+			}
+			if resp.Error != nil {
+				return fmt.Errorf("Error transferring %q: %v", t.Object.Oid, resp.Error)
+			}
+			if a.direction == Download {
+				// So we don't have to blindly trust external providers, check SHA
+				if err = tools.VerifyFileHash(t.Object.Oid, resp.Path); err != nil {
+					return fmt.Errorf("Downloaded file failed checks: %v", err)
+				}
+				// Move file to final location
+				if err = tools.RenameFileCopyPermissions(resp.Path, t.Path); err != nil {
+					return fmt.Errorf("Failed to copy downloaded file: %v", err)
+				}
+			} else if a.direction == Upload {
+				if err = api.VerifyUpload(t.Object); err != nil {
+					return err
+				}
+			}
+			wasAuthOk = true
+			complete = true
+		}
```

Comment:
This should handle the `default` case. A bad transfer adapter could cause Git LFS to hang, and then we'd get the bug report for it :)